### DISCUSSION TITLE
Write form data inside custom path in js

### DISCUSSION THREAD

**Discussion by lucafabbian** [pocketbase/pocketbase#5170] [Q&A] [answered]
I'm adding a custom route using javascript.
Is there a way to write to a custom path a file received inside a form?  If not, could it be implemented in some way?

```js

// read single multipart/form-data file
const doc = c.formFile("document")

// write it ?
$os.write("mycustomfolder/file.bin", doc, "666")

```

**Comment by ganigeorgiev** [comment, marked as answer]
It depends where you want the file to be stored and whether it would be associated to a record.

It is OK to use the `$os` utilities if you are targeting only the local file system but note that the method for file writes is [`$os.writeFile`](https://pocketbase.io/jsvm/functions/_os.writeFile.html). Additionally the content of the file must string/bytes.

If you want to use the PocketBase filesystem abstraction (aka. upload to S3 or local `pb_data/storage`) then you can do something like:

```js
let fsys;
try {
    fsys = $app.newFilesystem()

    fsys.uploadMultipart(c.formFile("document"), "your/file/key")

    // or manually create a $filesystem.File object
    // (it could be used also as part of the RecordUpsert form)
    const doc = $filesystem.fileFromMultipart(c.formFile("document"))
    fsys.uploadFile(doc, "your/file/key")
} finally {
    fsys?.close()
}
```

For uploading files to records programatically you can check the example in https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations (note that this will be simplified as part of the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355)).

_edit: wrapped the example in try/finally to ensure that fsys.close is invoked_

**Reply by lucafabbian** [reply]
Thanks, that's exactly what I was looking for.
